docs/wiki/ModulesDesign.md

```diff
@@ -124,33 +132,262 @@ name: '${deployment().name}-Table-${index}'
+
+```Bicep
+@description('Optional. The databases to create in the server')
+param databases array = []
+
+module server_databases 'databases/deploy.bicep' = [for (database, index) in databases: {}]
+```
+
+Each module should come with a `.bicep` folder with a least the `nested_cuaId.bicep` file in it
+
+### Patterns
+
+This sections shows you a few common patterns among resources that are usually very similar (e.g. providers)
+
+- [Locks](#locks)
+- [RBAC](#rbac)
+- [Diagnostic Settings](#diagnostic-settings)
+- [Private Endpoints](#private-endpoints)
+
+#### Locks
+
+The locks provider can be added as a `resource` to the resource template directly.
+
+```bicep
+@allowed([
+  'CanNotDelete'
+  'NotSpecified'
+  'ReadOnly'
+])
+@description('Optional. Specify the type of lock.')
+param lock string = 'NotSpecified'
+
+resource <mainResource>_lock 'Microsoft.Authorization/locks@2016-09-01' = if (lock != 'NotSpecified') {
+  name: '${<mainResource>.name}-${lock}-lock'
+  properties: {
+    level: lock
+    notes: (lock == 'CanNotDelete') ? 'Cannot delete resource or child resources.' : 'Cannot modify the resource or child resources.'
+  }
+  scope: <mainResource>
+}
+```
+
+### RBAC
+
+The RBAC deployment has 2 elements to it. A module that contains the implementation, and a module reference in the parent resource - each with it's own loop to enable you to deploy n-amount of role assignments to n-amount of principals.
+
+***1st Element in main resource***
+```bicep
+@description('Optional. Array of role assignment objects that contain the \'roleDefinitionIdOrName\' and \'principalId\' to define RBAC role assignments on this resource. In the roleDefinitionIdOrName attribute, you can provide either the display name of the role definition, or its fully qualified ID in the following format: \'/providers/Microsoft.Authorization/roleDefinitions/c2f4ef07-c644-48eb-af81-4b1b4947fb11\'')
+param roleAssignments array = []
+
+module <mainResource>_rbac '.bicep/nested_rbac.bicep' = [for (roleAssignment, index) in roleAssignments: {
+  name: '${deployment().name}-rbac-${index}'
+  params: {
+    principalIds: roleAssignment.principalIds
+    roleDefinitionIdOrName: roleAssignment.roleDefinitionIdOrName
+    resourceId: <mainResource>.id
+  }
+}]
+```
+
+***2nd Element as nested `.bicep/nested_rbac.bicep` file***
+
+Here you specify the platform roles available for the main resource. You can find further information in the [variables](#variables) section.
+
+The element requires you to provide both the `principalIds` & `roleDefinitionOrIdName` to assign to the principal IDs. Also, the `resourceId` is target resource's resource ID that allows us to reference it as an `existing` resource. Note, the implementation of the `split` in the resource reference becomes longer the deeper you go in the child-resource hierarchy.
```

Comment:
Should we add a link to the script that helps populate this list?

Reply:
Added a reference in the code itself